pythonforandroid/graph.py

```diff
@@ -99,17 +129,125 @@ def find_order(graph):
+    while len(to_be_added) > 0:
+        current_to_be_added = list(to_be_added)
+        to_be_added = []
+        for (added_tuple, adding_recipe) in current_to_be_added:
+            if len(added_tuple) > 1:
+                # No obvious commitment in what to add, ignore it.
+                # (Remember this function only catches obvious issues)
+                continue
+
+            name = added_tuple[0]
+            recipe_conflicts = []
+            recipe_dependencies = []
+            try:
+                # Get recipe to add and who's ultimately adding it:
+                recipe = Recipe.get_recipe(name, ctx)
+                recipe_conflicts = [c.lower() for c in recipe.conflicts]
+                recipe_dependencies = \
+                    get_dependency_tuple_list_for_recipe(
+                        recipe, blacklist=blacklist
+                    )
+            except ValueError:
+                pass
+            adder_recipe_1_name = adding_recipe or name
+
+            # See if this recipe itself conflicts with what was added before:
+            triggered_conflicts = []
+            for dep_tuple_list in deps:
+                if len(set(dep_tuple_list).intersection(
+                        set(recipe_conflicts))) == len(dep_tuple_list):
+                    triggered_conflicts.append(dep_tuple_list)
+
+            # See if what was added before conflicts with this recipe:
+            for dep_tuple_list in deps:
+                if len(dep_tuple_list) > 1:
+                    # Not an obvious commitment, so we won't check.
+                    # (remember this function only catches obvious issues)
+                    continue
+                try:
+                    conflicts = [c.lower() for c in
+                                 Recipe.get_recipe(dep_tuple_list[0], ctx).
+                                 conflicts]
+                    if name in conflicts:
+                        triggered_conflicts.append(dep_tuple_list)
+                except ValueError:
+                    pass
+
+            # Throw error on conflict:
+            if len(triggered_conflicts) > 0:
+                # Get first conflict and see who added that one:
+                adder_recipe_2_name = "'||'".join(triggered_conflicts[0])
+                adder_recipe_2_name =\
+                    (deps_were_added_by.get(adder_recipe_2_name,
+                                            adder_recipe_2_name) or
+                     adder_recipe_2_name)  # might be in dict, but None
+                # Prompt error:
+                raise BuildInterruptingException(
+                    "Conflict detected:" +
+                    " '" + str(adder_recipe_1_name) + "'" +
+                    " inducing dependencies " + str((recipe.name,)) +
+                    " and '" + str(adder_recipe_2_name) + "'" +
+                    " inducing conflicting dependencies " +
+                    str(triggered_conflicts[0])
+                )
```

Comment:
Minor: isn't `format()` producing more beautiful/readable code or is it more a matter of taste?

Reply:
FWIW, my personal preference is`format()`. `format` is a nice migration path to fstrings (the *most* readable option), for the day when python 3.5 and earlier support is dropped.

If there are too many arguments to format to make it clear what is what, I'll sometimes pass an unpacked dictionary (e.g. `**format_opts`)and use the key names explicitly, e.g.   `'and {key1} inducing conlicting dependencies {key2}'.format(**fmt_opts)`

Reply:
yea it should be format. was just me farting on my keyboard, there is no good reason it looks ugly like that

Reply:
I used format now! are you happy with how it looks like, or should I use a dictionary with parameter names?

Reply:
Thanks :+1: